AssemblyProvider.cs

```diff
@@ -0,0 +1,274 @@
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using AsmResolver.PE.File;
+using AsmResolver.PE.File.Headers;
+using ELFSharp.ELF;
+using ELFSharp.ELF.Sections;
+using MachOSharp;
+using MachOSharp.Command;
+using MachOSharp.Util;
+using SharpPdb.Native;
+using ElFMachine = ELFSharp.ELF.Machine;
+
+namespace FactorioAchievementPatcher;
+
+public interface AssemblyProvider : IDisposable {
+
+    public static AssemblyProvider Create(string modulePath, byte[] moduleBytes) {
+        if (Path.GetExtension(modulePath).Equals(".exe")) {
+            return new WindowsAssemblyProvider(modulePath);
+        }
+
+        if (ELFReader.TryLoad(new MemoryStream(moduleBytes), true, out ELF<ulong> elf)) {
+            return new LinuxAssemblyProvider(elf);
+        }
+
+        if (MachOReader.TryLoadFat(new MemoryStream(moduleBytes), out var fat, false)) {
```

Comment:
`fat` is not disposable here?

Reply:
Nope, MachOFat doesn't keep the stream open, or use any lazy loading.